go/consensus/tendermint/abci/system.go

```diff
@@ -0,0 +1,122 @@
+)
+
+// prepareSystemTxs prepares a list of system transactions to be included in a proposed block in
+// case where the node is currently the block proposer.
+func (mux *abciMux) prepareSystemTxs() [][]byte {
+	var systemTxs [][]byte
+
+	// Append block metadata as a system transaction.
+	stateRoot, err := mux.state.workingStateRoot()
+	if err != nil {
+		panic(fmt.Errorf("failed to compute working state root: %w", err))
+	}
+
+	blockMeta := consensus.NewBlockMetadataTx(&consensus.BlockMetadata{
+		StateRoot: stateRoot,
+	})
+	sigBlockMeta, err := transaction.Sign(mux.state.identity.ConsensusSigner, blockMeta)
+	if err != nil {
+		panic(fmt.Errorf("failed to sign block metadata transaction: %w", err))
+	}
+	systemTxs = append(systemTxs, cbor.Marshal(sigBlockMeta))
+	mux.state.blockCtx.SystemTransactions = append(mux.state.blockCtx.SystemTransactions, blockMeta)
+	mux.state.proposal.resultsDeliverTx = append(mux.state.proposal.resultsDeliverTx, &types.ResponseDeliverTx{
+		Code: types.CodeTypeOK,
+		Data: cbor.Marshal(nil),
+	})
+
+	return systemTxs
+}
+
+// processSystemTx processes a system transaction in DeliverTx context.
+func (mux *abciMux) processSystemTx(ctx *api.Context, tx *transaction.Transaction) error {
+	if ctx.Mode() != api.ContextDeliverTx {
+		return fmt.Errorf("system methods are not allowed to be called")
+	}
+
+	// NOTE: The panics below will either trigger a bad proposal being rejected (on nodes in the
+	//       validator set) or a failure in block processing (on other nodes).
+
+	// Since system transactions can only be injected by the proposer, make sure they are not
+	// themselves included in the proposal phase.
+	if len(mux.state.proposal.hash) == 0 {
+		panic(fmt.Errorf("system transaction included during proposal phase"))
+	}
+
+	// Nonce must be zero and fee must be nil.
+	if tx.Nonce != 0 || tx.Fee != nil {
+		panic(fmt.Errorf("malformed system transaction in block"))
+	}
+	// Transaction must be signed by the proposer.
+	txSigner := ctx.TxSigner()
+	txSignerAddress := []byte(tmcrypto.PublicKeyToTendermint(&txSigner).Address())
+	if proposerAddress := ctx.BlockContext().ProposerAddress; !bytes.Equal(txSignerAddress, proposerAddress) {
+		panic(fmt.Errorf("system transaction not signed by block proposer (expected: %x got: %x)", proposerAddress, txSignerAddress))
+	}
+
+	// Accumulate system transactions for later verification.
+	ctx.BlockContext().SystemTransactions = append(ctx.BlockContext().SystemTransactions, tx)
+
+	return nil
+}
+
+// validateSystemTxs performs system transaction validation at the end of a block.
+func (mux *abciMux) validateSystemTxs() error {
+	// Don't perform any validation when we are still building the proposal.
+	if len(mux.state.proposal.hash) == 0 {
+		return nil
+	}
+
+	var hasBlockMetadata bool
+	for _, tx := range mux.state.blockCtx.SystemTransactions {
+		switch tx.Method {
+		case consensus.MethodMeta:
+			// Block metadata, verify state root.
+			if hasBlockMetadata {
```

Comment:
Having a counter instead of a bool flag could remove this `if`. Not sure what is better :)